### PR TITLE
Anthony  toggle show password

### PR DESCRIPTION
**Pull request by anthony-hull**
### Details
Added a toggle button icon to password fields to change visibility

### Fixed Issues
https://github.com/Expensify/App/issues/2734

### QA Steps
Web and Desktop:
1. Type something into a password field
2. Verify password is hidden
2. Click eye icon
3. Verify the password is visible
4. Click again
5. Verify it is hidden again

Android and iOS
1. Type something into a password field
2. Verify password is shown
2. Click eye icon
3. Verify the password is hidden
4. Click again
5. Verify it is shown again

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
setpassword screen web:
https://user-images.githubusercontent.com/47184118/129957727-cff28fb8-f9f5-4d6c-9d69-cd469e2e74e7.mp4

#### Edge
image
note the lack of vendor specific show password functionality

#### Mobile Web
image

#### Desktop
https://user-images.githubusercontent.com/47184118/133485179-aa45ff31-6457-4865-a207-6eff4177ae66.mov

#### iOS
image

#### Android
image

### CODE REVIEW

**Comment by anthony-hull** [comment]
I have yet to upload all the proof of testing as I have an outstanding implementation question in the PR

**Comment by johnmlee101** [comment]
Where is the question?

**Comment by anthony-hull** [comment]
it is inside the review I started. you need to click view changes:
image

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:184]: @johnmlee101

**Comment by anthony-hull** [comment]
@johnmlee101 can you see my review now?

**Comment by johnmlee101** [comment]
Yes!

**Comment by anthony-hull** [comment]
I just have to fix what broke in the merge

**Comment by parasharrajat** [comment]
I noticed that in the Edge browser Eye icon is already shown to reveal the password. Are you going to disable that? Otherwise, two icons will be shown.

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (parasharrajat) [inline src/components/ExpensiTextInput/BaseExpensiTextInput.js:188]: I feel like this part can go into a separate component which only deals with PasswordInput. I would love to hear your thoughts.

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (parasharrajat) [inline src/components/ExpensiTextInput/BaseExpensiTextInput.js:145]: Hmm, Did you mistakenly moved the error message to the top of the Input?

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:144]: probably. git made a complete mess of this when merging during rebase

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:187]: I considered this.
Do you mean a wrapper secure text component which passes in the icon control via props to this component?

Comment (parasharrajat) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:187]: So something like called as `PasswordExpensiInput`.

```javascript
<Wrapper>
  <ExpensiInput secureInput/>
  <icon>
</Wrapper>
```

Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:187]: But the icon needs to be inside the layout of the input as it needs to take up space inside the flow of the flex container to displace any icons added inside the input by password managers.

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js]: it was duplicated rather than moved

**Comment by anthony-hull** [comment]
@johnmlee101 can you please review this PR?

**Comment by johnmlee101** [comment]
As mentioned this will be on hold, but I'll still review!

File: src/styles/styles.js
Comment (johnmlee101) [inline src/styles/styles.js:516]: Are these values set in stone? What happens if its set as our variables `componentBorderRadiusNormal`, etc?

**Review by johnmlee101 — CHANGES_REQUESTED** [review]
You have an empty file that can be deleted 
`src/components/ExpensiTextInput/disableEdgeIcon.css`

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:40]: I considered this but it is already a prop as part of react native, the interface hasn't changed. Do we still need to document it?

Comment (johnmlee101) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:40]: Oh I see, don't need to then.

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:146]: It was that way when I started the branch. This is an incoming change that got reverted in the rebase:
image
Do you want to keep the incoming style change or for me to revert it as this PR stands?

File: src/styles/styles.js
Comment (anthony-hull) [reply src/styles/styles.js]: I've removed these

File: src/styles/styles.js
Comment (anthony-hull) [reply src/styles/styles.js:476]: So far I can't see any regressions in the usages of the inputs across the app.
I'll do another pass since I cleaned up the CSS a little

Comment (johnmlee101) [reply src/styles/styles.js:476]: Cool, that would be much appreciated!

**Comment by anthony-hull** [comment]
there are some bugs introduced in the input field after merging main:
https://expensify.slack.com/archives/C01GTK53T8Q/p1633367678385600?thread_ts=1633367592.384900&cid=C01GTK53T8Q
It was working on all platforms before merging

**Comment by anthony-hull** [comment]
https://github.com/Expensify/App/issues/5632
this is the related issue

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js]: What is the right course of action for this in future? I know style changes are not usually merged. But here a style change slipped in and is a change I had to merge or reject when rebasing.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (anthony-hull) [reply src/components/ExpensiTextInput/BaseExpensiTextInput.js:226]: is it to stop a new anon function being created in each render?

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Although the code looks good. But I have found a few issues during testing.

Check the Padding on the right of input 
image

Staging 
image


---
Check the Label.
image

---
Looks like you are adding opposite behavior.
How it should be?
Screenshot 2021-10-26 04:02:34

How it is now.
image

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (parasharrajat) [inline src/components/ExpensiTextInput/BaseExpensiTextInput.js:203]: This PR will bring back the issue that we recently fixed https://github.com/Expensify/App/issues/4515.
When you click on the highlighted area
issue

File: src/styles/styles.js
Comment (parasharrajat) [inline src/styles/styles.js:570]: Why width 90%?

File: src/styles/styles.js
Comment (parasharrajat) [inline src/styles/styles.js:585]: I didn't feel good about using the margin. Can we achieve this using padding?

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Sorry for the delayed review. I got busy recently with family stuff. 

Nice, you fixed the Overflow issue. But there is still an issue. 

Apart from it, the row containing the icon input has vertical padding which is overflowing the input container. It does not have a visual issue but breaks the box model and could cause issues in the future. 

I noticed that you removed the H-padding from the input which is going to conflict with my upcoming PR https://github.com/Expensify/App/pull/5805. 

Anyways, whenever you fix the below-mentioned issue. it would be resolved as well.

Sorry, for stretching the review. I am trying to remove as many as regressions. Thanks for your understanding and time.

**Comment by anthony-hull** [comment]
don't need to apologise, family come first 👍 I appreciate the feedback as well. 

I need to push the input element downwards but keep the other element in the row in place so I will use something apart form padding to displace it down.

I'll look at the PR you linked and fix it. I'm going on holiday today so I won't get back to this until the 10th Nov.

**Comment by parasharrajat** [comment]
Any update on this @anthony-hull . I can say you are close.

**Comment by anthony-hull** [comment]
I'm working on this over the next 24 hours

**Comment by anthony-hull** [comment]
image
image
image


vs 

image
image

About the spacing between the input and the icon, what do we think?

**Comment by parasharrajat** [comment]
Spacing should symmetrical around the icon. So that icon would fit in a rect. 
image
But equal space around the icon on x axis.

**Comment by anthony-hull** [comment]
in the image you chose the spacing between the input element and the eye is 11.5px and the spacing between the border and the eye is also 11.5px

**Comment by parasharrajat** [comment]
I was thinking to make it so that the icon looks like a square. But we need more eyes cc: @shawnborton

**Comment by shawnborton** [comment]
Hmm I'm a little lost regarding what's being asked of me here, what do you mean more of a square?

**Comment by parasharrajat** [comment]
I was thinking if we increase the space around the icon horizontally then it will look better. But I tagged you to ask about this https://github.com/Expensify/App/pull/4746#issuecomment-978071261 @shawnborton

**Comment by shawnborton** [comment]
Ah, got it. I like the mockups where there is more space between the 1Pass icon and the eye. Maybe we can just use 8px gap between those two icons and see how it looks?

**Comment by anthony-hull** [comment]
image
image

**Comment by shawnborton** [comment]
Let's go with the top version.

**Comment by anthony-hull** [comment]
Just need to test on all platforms  but something has come up this evening. I'll be able to carry on tomorrow

**Comment by parasharrajat** [comment]
Is this ready for review?

**Comment by anthony-hull** [comment]
It works on Web, Mobile Web and Desktop. I can't test on the native apps at the moment as they crash due to another issue

**Comment by parasharrajat** [comment]
The crash is fixed now. You can pull new changes and test.

**Comment by anthony-hull** [comment]
oh man tired of fixing breaking changes from merges 😆

**Comment by anthony-hull** [comment]
@parasharrajat testing complete, please review

**Comment by parasharrajat** [comment]
I hate to say this but overflow issue is back 
image

File: src/styles/styles.js
Comment (parasharrajat) [inline src/styles/styles.js:651]: Well, this margin seems off. Why do we need it?

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (parasharrajat) [inline src/components/ExpensiTextInput/BaseExpensiTextInput.js:12]: Could you please rename it to `Expensicons` instead of `icons`.

File: src/components/ExpensiTextInput/BaseExpensiTextInput.js
Comment (parasharrajat) [inline src/components/ExpensiTextInput/BaseExpensiTextInput.js:219]: We can get rid of this line...please

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Just an overflow issue again and a few tweaks. I think that the current implementation could be improved by using an absolute positioned icon over the input. But as this is working fine, I won't ask you to do that unless you are willing to do that change.

File: src/styles/styles.js
Comment (anthony-hull) [reply src/styles/styles.js:651]: it's to reduce the distance between the icon and the input element to be 8px.
I decided to do it with a negative margin so I didn't have to add any JS logic to the stylings based of if the icon was there or not. I generally prefer to do things in CSS if I can without adding JS.

**Comment by anthony-hull** [comment]
> Just an overflow issue again and a few tweaks. I think that the current implementation could be improved by using an absolute positioned icon over the input. But as this is working fine, I won't ask you to do that unless you are willing to do that change.

If it were absolute, wouldn't any password manager icons render underneath it? as the input element would fill the space due to the icon being out of flow?

File: src/styles/styles.js
Comment (parasharrajat) [reply src/styles/styles.js:651]: Well, the negative margin has side-effect on Android and thus we should not use it on the Clickable element. https://reactnative.dev/docs/handling-touches#known-issues

Anyways, the whole app is built on js so we should not worry about using a flag to adjust the style.

Comment (anthony-hull) [reply src/styles/styles.js:651]: can't argue with not supported!

**Comment by parasharrajat** [comment]
> If it were absolute, wouldn't any password manager icons render underneath it? as the input element would fill the space due to the icon being out of flow?

Good point. Then we are good.

**Comment by parasharrajat** [comment]
Hey, @anthony-hull I forgot to mention earlier that you have to sign all of your commits before we can merge the PR.

**Comment by parasharrajat** [comment]
@anthony-hull  Any update on the requested changes. Let's complete this asap. Let me know if you are facing any challenges.

**Comment by anthony-hull** [comment]
I've fixed the overflowing input. I'll sign the commits, re-implement the negative margin and then re-test. should be ready soon!

**Comment by anthony-hull** [comment]
do I have to have all of them to show as verified? I ran a rebase command to sign them all and I was having fix merge issues on all my changes again. git was not happy. it was a nightmare so I aborted.

`git rebase --exec 'git commit --amend --no-edit -n -S' -i main`

**Comment by parasharrajat** [comment]
Unfortunately Yes.

**Comment by anthony-hull** [comment]
much easier to make a new PR with a new branch :)
please see:
https://github.com/Expensify/App/pull/6501
 I just need to test on all platforms now. I'll let you know!

**Comment by parasharrajat** [comment]
@anthony-hull Please close this PR. Thanks.

**Pull request closed without merging**
